### PR TITLE
[AIRFLOW-5413] Refactor worker config

### PR DESCRIPTION
**Pull request by davlum**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-5413) issues and references them in the PR title. 

### Description

- [X] This PR enables users of the K8S API to pass in pod configuration as a string, or as a path to a JSON/YAML file.

### Tests

- [X] My PR adds the following unit tests.

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by dimberman** [comment]
@ashb I asked @davlum to separate any infra changes from API changes so we can have an easier time reviewing anything front facing.

File: airflow/kubernetes/pod_generator.py [hunk 1]
```diff
@@ -290,51 +338,120 @@ def reconcile_pods(base_pod: k8s.V1Pod, client_pod: k8s.V1Pod) -> k8s.V1Pod:
...
-        merge_objects(base_pod, client_pod_cp)
+            setattr(client_obj_cp, field_name, appended_fields)
+            return client_obj_cp
+
+        if base_pod.spec and not client_pod.spec:
+            client_pod_cp.spec = base_pod.spec
+        elif client_pod_cp.spec and base_pod.spec:
+            client_container = client_pod_cp.spec.containers[0]
+            base_container = base_pod.spec.containers[0]
+            cc1 = extend_object_field(base_container, client_container, 'volume_mounts')
```
Comment (dimberman) [inline airflow/kubernetes/pod_generator.py]: Why are you creating all these temp vars? This looks concerningly scala-esque.

**Review by dimberman — CHANGES_REQUESTED** [review]

File: airflow/kubernetes/pod_generator.py (hunk 1 above)
Comment (davlum) [reply airflow/kubernetes/pod_generator.py]: I could use the same variable name.

Comment (dimberman) [reply airflow/kubernetes/pod_generator.py]: Yeah using the same variable would make sense here. You don't need any of the middle ones.

**Comment by davlum** [comment]
I'm finding the `reconcile_pods` function to be a bit large and unwieldy. I'm thinking of splitting it out into `reconcile_spec`, `reconcile_metadata`, `reconcile_containers` etc. and then having them all called by `reconcile_pods`. It will also be easier to test the individual components. Thoughts?

**Comment by ashb** [comment]
> I'm finding the `reconcile_pods` function to be a bit large and unwieldy. I'm thinking of splitting it out into `reconcile_spec`, `reconcile_metadata`, `reconcile_containers` etc. and then having them all called by `reconcile_pods`. It will also be easier to test the individual components. Thoughts?

Within reason, smaller functions that do one thing are better, yes.

**Review by dimberman — APPROVED** [review]
@davlum this looks fantastic great job :).

**Comment by kaxil** [comment]
Something is up with travis, it is building your old commit, can you squash all commits into 1 @davlum

**Comment by davlum** [comment]
@kaxil I squashed to one commit and pushed. It seems to trigger two travis builds, one that corresponds to the [correct version of the code](https://travis-ci.org/davlum/incubator-airflow?utm_medium=notification&utm_source=github_status), but isn't linked to the PR and one that is linked to the PR, but corresponds to an older version of the code. Oddly, they both claim to link to the same commit. Very buggy behaviour from Travis.

**Pull request closed without merging**